doc/design/programming_solver.md

```diff
@@ -0,0 +1,280 @@
+
+You can learn from the examples in the above links that:
+
+- Using Matlab and R are quite the same, they require users to define the constraints as matrixes and call a function to get the result. They both have their own grammar, and you have to write code according to their language specifications.
+- Using "AMPL like" high-level language to describe the problem, and call the corresponding solvers. Pyomo and AMPL have similar components in their grammar: sets, parameters, variables, objectives, constraints (https://pyomo.readthedocs.io/en/stable/pyomo_modeling_components/index.html).
+
+So we can see that using AMPL is a modern and general way of describing mathematical programming problems. We can simply write the AMPL snippet to describe the above problem:
+
+```
+set X;
+var ProductAmt{x in X} >= 0;
+
+param Price{x in X};
+param MaterialCost{x in X};
+param OtherCost{x in X};
+param Finishing{x in X};
+param Carpentry{x in X};
+param Demand{x in X};
+
+maximize profit:
+    sum{x in X} (Price[x] - MaterialCost[x] - OtherCost[x]) * ProductAmt[x];
+
+s.t. finishing: sum{x in X} Finishing[x] * ProductAmt[x] <= 100;
+s.t. carpentry: sum{x in X} Carpentry[x] * ProductAmt[x] <= 80;
+s.t. demand{x in X}: ProductAmt[x] <= 40;
+```
+
+## Grammar Design
+
+In order to extend SQL to have completely same ability of AMPL, the extended syntax should be able to describe **objective and constraints** while the input data table can store the **params** for each **variable**, and the rows in the table is naturally become the **set** we defined in AMPL. 
+
+### Single Variable
+
+Then we have the below `train_table`:
+
+| product | price | materials_cost | other_cost | finishing | carpentry | max_num |
+| ------- | ----- | -------------- | ---------- | --------- | --------- | ------- |
+| soldier | 27    | 10             | 14         | 2         | 1         | 40      |
+| train   | 21    | 9              | 10         | 1         | 1         | 10000   |
+
+In the `train_table`:
+
+- The set X is row one and row two.
+- We have one variable, and the variable name strings is stored in column `product`. In cases that have multiple variables (like the example described at https://en.wikipedia.org/wiki/AMPL), the table should have multiple string columns to store the variable.
+- Other columns like `price`, `materials_cost` are all params for the corresponding variable.
+
+Then we can use below extended SQL syntax to describe above example:
+
+```sql
+SELECT * FROM train_table
+TO MAXIMIZE SUM((price - materials_cost - other_cost) * product)
+CONSTRAINT SUM(finishing * product) <= 100
+CONSTRAINT SUM(carpentry * product) <= 80
+CONSTRAINT product <= max_num
+WITH variables="product",
+     var_type="product:Integers"
```

Comment:
I like this design very much. It is a breakthrough to realize that the data schema is also the structure of the optimization problem, which I mean that for this problem, we can merge the parameter table together with the result table because they have corresponding rows:

```
| product | price | material_cost | other_cost | finishing | carpentry | max_num | number/result |
```

Given this property, I guess we might not need 

```
WITH varaibles="product", var_type="product:Intergers"
```

Instead, as long as we specify the data type of the column `result` in the above table having SQL data type `INT`, we can write

```
WITH result
```

And we get 

```sql
SELECT * FROM woodcarving
TO MAXIMIZE SUM((price - materials_cost - other_cost) * product)
WITH result  # WITH specify the column of variables, which is also the output.
CONSTRAINT SUM(finishing * product) <= 100
CONSTRAINT SUM(carpentry * product) <= 80
CONSTRAINT product <= max_num;
```

Reply:
Well, I was considering exactly the same thing earlier, yet I found it may be a bit hard to implement:

1. Hive and MaxCompute do not support `UPDATE` statement as a normal DBMS does. Even though they support `INSERT OVERWRITE`( see https://www.revisitclass.com/hadoop/insert-overwrite-table-values-in-hive/, and https://yq.aliyun.com/articles/66890), it still works completely different from `UPDATE` which may change the original table data order or interference other concurrent jobs.
1. Not all cases using a table as the input ( please take the multivariable case as an example), and I think we should be able to support the user write a complicated SQL before `TO MAXIMIZE|MINIMIZE`.
1. The user have to write statements to create a result column even if he/she have a handy table stores the variables and parameters.

Reply:
I see. Then can we use `WITH product=INT` than `WITH variable=... var_type=...`?

Reply:
Yes, sounds good.